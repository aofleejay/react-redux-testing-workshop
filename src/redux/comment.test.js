import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import axios from 'axios'
import commentReducer, { receiveComments, getComments } from './comment'

describe('Test comment actions', () => {
  const comments = [
    { title: 'title1', body: 'body1', author: 'author1', postat: '1m' },
    { title: 'title2', body: 'body2', author: 'author2', postat: '2m' },
  ]

  afterEach(() => {
    jest.restoreAllMocks()
  })

  it('receiveComments action creator return correct action', () => {
    const expected = { type: 'GET_COMMENTS_SUCCESS', comments: comments }
  
    const actual = receiveComments(comments)
  
    expect(actual).toEqual(expected)
  })

  it('getComments dispatch action correctly', () => {
    const mockStore = configureMockStore([thunk])
    const store = mockStore({ comments: [] })
    const mockGet = jest.spyOn(axios, 'get').mockResolvedValue({ data: comments })

    store.dispatch(getComments()).then(() => {
      expect(store.getActions()).toEqual([{ type: 'GET_COMMENTS_SUCCESS', comments: comments }])
    })
  })
})

describe('Test comment reducer', () => {
  const initialState = []

  it('Should return new comment list when action type is GET_COMMENTS_SUCCESS', () => {
    const action = {
      type: 'GET_COMMENTS_SUCCESS',
      comments: [
        { title: 'title1', body: 'body1' }
      ]
    }
    const expected = [
      { title: 'title1', body: 'body1' }
    ]
    
    const actual = commentReducer(initialState, action)

    expect(actual).toEqual(expected)
  })

  it('Should return empty array when action type is GET_COMMENTS_FAILED', () => {
    const action = {
      type: 'GET_COMMENT_FAILED',
    }
    const expected = []

    const actual = commentReducer(initialState, action)

    expect(actual).toEqual(expected)
  })

  it('Should return initial state when action type is OTHER_ACTION_TYPE', () => {
    const action = {
      type: 'OTHER_ACTION_TYPE',
    }

    const actual = commentReducer(initialState, action)

    expect(actual).toEqual(initialState)
  })
})
