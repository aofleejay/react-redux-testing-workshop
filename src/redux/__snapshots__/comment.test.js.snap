// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test comment actions Test getComments asychronous action creator Should dispatch GET_COMMENTS_FAILED action when promise rejected 1`] = `
Array [
  Object {
    "type": "GET_COMMENTS_FAILED",
  },
]
`;

exports[`Test comment actions Test getComments asychronous action creator Should dispatch GET_COMMENTS_SUCCESS action when promise resovled 1`] = `
Array [
  Object {
    "comments": Array [
      Object {
        "author": "author1",
        "body": "body1",
        "postat": "1m",
        "title": "title1",
      },
      Object {
        "author": "author2",
        "body": "body2",
        "postat": "2m",
        "title": "title2",
      },
    ],
    "type": "GET_COMMENTS_SUCCESS",
  },
]
`;

exports[`Test comment actions Test receiveComments action creator Should return GET_COMMENTS_SUCCESS action 1`] = `
Object {
  "comments": Array [
    Object {
      "author": "author1",
      "body": "body1",
      "postat": "1m",
      "title": "title1",
    },
    Object {
      "author": "author2",
      "body": "body2",
      "postat": "2m",
      "title": "title2",
    },
  ],
  "type": "GET_COMMENTS_SUCCESS",
}
`;

exports[`Test comment reducer Should return empty array when action type is GET_COMMENTS_FAILED 1`] = `Array []`;

exports[`Test comment reducer Should return initial state when action type is OTHER_ACTION_TYPE 1`] = `Array []`;

exports[`Test comment reducer Should return new comment list when action type is GET_COMMENTS_SUCCESS 1`] = `
Array [
  Object {
    "body": "body1",
    "title": "title1",
  },
]
`;
